staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -439,12 +428,23 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 //           |                                                       |
 //  AfterShutdownDelayDuration (delayedStopCh)   PreShutdownHooksStopped (preShutdownHooksHasStoppedCh)
 //           |                                                       |
+//           |-------------------------------------------------------|
+//                                    |
+//                                    |
+//               NotAcceptingNewRequest (notAcceptingNewRequestCh)
+//                                    |
+//                                    |
+//           |-------------------------------------------------------|
+//           |                                                       |
+//           |                                                       |
 //           |----------------------------------                     |
 //           |                                  |                    |
 //           |                       (HandlerChainWaitGroup::Wait)   |
 //           |                                  |                    |
 //           |                  InFlightRequestsDrained (drainedCh)  |
 //           |                                  |                    |
+//           |                       (AuditBackend::Shutdown())      |
```

Comment:
Actually, this isn't correct any more (and in fact it wasn't fully correct before too)

You should switch to
```
//               NotAcceptingNewRequest (notAcceptingNewRequestCh)
//                                    |
//                                    |
//           |---------------------------------------------------------|
//           |                        |              |                 |
//        [without                 [with             |                 |
// ShutdownSendRetryAfter]  ShutdownSendRetryAfter]  |                 |
//           |                        |              |                 |
//           |                        ---------------|                 |
//           |                                       |                 |
//           |                         (HandlerChainWaitGroup::Wait)   |
//           |                                       |                 |
//           |                    InFlightRequestsDrained (drainedCh)  |
//           |                                       |                 |
//           ----------------------------------------|-----------------|
//                                 |                 |
//                         stopHttpServerCh (AuditBackend::Shutdown())
//                                 |
```
...


I need to verify that closer, but the above is definitely closer to the thruth.

Reply:
OK - I think the above is correct:
- handlerChain + inflight requests drained - is happening unconditionally always
- without ShutdownSendRetryAfter - we don't block on the above - just immediately go to stopHttpServerCh
- with ShutdownSendRetryAfter - we actually block on drainedCh instead
- AuditBackend::shutdown is always blocked on drainedCh

The above diagram looks correct to me and the code reflects it. So once fixed this LGTM.

Reply:
yes, it's correct, for someone reading it the first time, i think it's going to be hard to follow with both mode mingled together. I would recommend having separate diagrams for each mode.